lib/plugins/main.js

```diff
@@ -1,361 +1,141 @@
-        options.services.messaging.fcmPublicVapidKey
-      ) { %>
-        fireMess.usePublicVapidKey(<%=
-          serialize(options.services.messaging.fcmPublicVapidKey)
-        %>)
-      <% } %>
-
-      inject('fireMess', fireMess)
-      inject('fireMessObj', fireMessObj)
-    }
+    Array.prototype.push.apply(servicePromises, [
+      <%= clientServices.reduce((acc, service) => `${acc}${service}Service(session, firebase),\n      `, '') %>
+    ])
   }
-
   <% } %>
 
-  <% if (options.services.performance) { %>
-  /** --------------------------------------------------------------------------------------------- **/
-  /** ----------------------------------- FIREBASE PERFORMANCE ------------------------------------ **/
-  /** --------------------------------------------------------------------------------------------- **/
-
-  // Firebase Performance can only be initiated on client side
-  if(process.client) {
-    <%= writeImportStatement('performance') %>
-
-    const firePerf = session.performance()
-    const firePerfObj = firebase.performance
-    inject('firePerf', firePerf)
-    inject('firePerfObj', firePerfObj)
-  }
+  <% /* No need to use process.client here, all client stuff will just be undefined on server */ %>
+  const [
+    <%= serverServices.map(getServiceName).join(',\n    ') + (clientServices.length ? ',' : '') %>
+    <%= clientServices.map(getServiceName).join(',\n    ') %>
+  ] = await Promise.all(servicePromises)
 
+  <% if (options.legacyMode) { %>
+  <% for (const service of serverServices) { %>
+  inject('fire<%= options.injectionMapping[service] %>', <%= getServiceName(service) %>)
+  inject('fire<%= options.injectionMapping[service] %>Obj', firebase.<%= service %>)
   <% } %>
 
-  <% if (options.services.analytics) { %>
-  /** --------------------------------------------------------------------------------------------- **/
-  /** ----------------------------------- FIREBASE ANALYTICS -------------------------------------- **/
-  /** --------------------------------------------------------------------------------------------- **/
-
-  // Firebase Analytics can only be initiated on the client side
   if (process.client) {
-    <%= writeImportStatement('analytics') %>
-
-    const fireAnalytics = session.analytics()
-    const fireAnalyticsObj = firebase.analytics
-    inject('fireAnalytics', fireAnalytics)
-    inject('fireAnalyticsObj', fireAnalyticsObj)
-
-    <% if (typeof options.services.analytics === 'object') { %>
-
-    <% if ('collectionEnabled' in options.services.analytics) { %>
-
-    // In development we want to disable analytics collection
-    fireAnalytics.setAnalyticsCollectionEnabled(<%= !!options.services.analytics.collectionEnabled %>)
-
-    <% } %>
-
+    <% for (const service of clientServices) { %>
+    inject('fire<%= options.injectionMapping[service] %>', <%= getServiceName(service) %>)
+    inject('fire<%= options.injectionMapping[service] %>Obj', firebase.<%= service %>)
```

Comment:
```suggestion
    inject('fire<%= options.injectionMapping[service] %>Obj', firebase.<%= options.serviceMapping[service] %>)
```

Same as above. Does't make a difference here because the keys and the values of the clientServices are exactly the same in the serviceMapping, but that's just luck.